experimental/wazerotest/wazerotest.go

```diff
@@ -0,0 +1,605 @@
+}
+
+func (m *Module) String() string {
+	return "module[" + m.ModuleName + "]"
+}
+
+func (m *Module) Name() string {
+	return m.ModuleName
+}
+
+func (m *Module) Memory() api.Memory {
+	if m.ExportMemory != nil {
+		m.once.Do(m.initialize)
+		return m.ExportMemory
+	}
+	return nil
+}
+
+func (m *Module) ExportedFunction(name string) api.Function {
+	m.once.Do(m.initialize)
+	return m.exportedFunctions[name]
+}
+
+func (m *Module) ExportedFunctionDefinitions() map[string]api.FunctionDefinition {
+	m.once.Do(m.initialize)
+	return m.exportedFunctionDefinitions
+}
+
+func (m *Module) ExportedMemory(name string) api.Memory {
+	if m.ExportMemory != nil && name == "memory" {
+		m.once.Do(m.initialize)
+		return m.ExportMemory
+	}
+	return nil
+}
+
+func (m *Module) ExportedMemoryDefinitions() map[string]api.MemoryDefinition {
+	m.once.Do(m.initialize)
+	return m.exportedMemoryDefinitions
+}
+
+func (m *Module) ExportedGlobal(name string) api.Global {
+	m.once.Do(m.initialize)
+	return m.exportedGlobals[name]
+}
+
+func (m *Module) NumFunction() int {
+	return len(m.Functions)
+}
+
+func (m *Module) Function(i int) api.Function {
+	m.once.Do(m.initialize)
+	return m.Functions[i]
+}
+
+func (m *Module) NumGlobal() int {
+	return len(m.Globals)
+}
+
+func (m *Module) Global(i int) api.Global {
+	m.once.Do(m.initialize)
+	return m.Globals[i]
+}
+
+func (m *Module) Close(ctx context.Context) error {
+	return m.CloseWithExitCode(ctx, 0)
+}
+
+func (m *Module) CloseWithExitCode(ctx context.Context, exitCode uint32) error {
+	atomic.CompareAndSwapUint64(&m.exitStatus, 0, exitStatusMarker|uint64(exitCode))
+	return nil
+}
+
+func (m *Module) ExitStatus() (exitCode uint32, exited bool) {
+	exitStatus := atomic.LoadUint64(&m.exitStatus)
+	return uint32(exitStatus), exitStatus != 0
+}
+
+func (m *Module) initialize() {
+	m.exportedFunctions = make(map[string]api.Function)
+	m.exportedFunctionDefinitions = make(map[string]api.FunctionDefinition)
+	m.exportedGlobals = make(map[string]api.Global)
+	m.exportedMemoryDefinitions = make(map[string]api.MemoryDefinition)
+
+	for index, function := range m.Functions {
+		for _, exportName := range function.ExportNames {
+			m.exportedFunctions[exportName] = function
+			m.exportedFunctionDefinitions[exportName] = function.Definition()
+		}
+		function.module = m
+		function.index = index
+	}
+
+	for _, global := range m.Globals {
+		for _, exportName := range global.ExportNames {
+			m.exportedGlobals[exportName] = global
+		}
+	}
+
+	if m.ExportMemory != nil {
+		m.ExportMemory.module = m
+		m.exportedMemoryDefinitions["memory"] = m.ExportMemory.Definition()
+	}
+}
+
+// Global is an implementation of the api.Global interface, it represents a
+// global in a WebAssembly module.
+type Global struct {
+	internalapi.WazeroOnlyType
+
+	// Type of the global value, used to interpret bits of the Value field.
+	ValueType api.ValueType
+
+	// Value of the global packed in a 64 bits field.
+	Value uint64
```

Comment:
works for most things except v128 iirc

Reply:
We don't expose v128 in `api.ValueType` yet, I would suggest we keep it as-is for now and if we eventually expose a constant for the 128 bit vector we can revisit this API as well.